subsys/input/input_lvgl.c

```diff
@@ -0,0 +1,251 @@
+		data->pending_event.key = evt->code;
+		break;
+	case LV_INDEV_TYPE_BUTTON:
+		data->pending_event.btn_id = evt->code;
+		data->pending_event.state = evt->value ? LV_INDEV_STATE_PR : LV_INDEV_STATE_REL;
+		break;
+	case LV_INDEV_TYPE_ENCODER:
+		data->pending_event.enc_diff = evt->value;
+		break;
+	default:
+		LOG_ERR("Received callback for unknown indev: %d", data->indev_drv.type);
+	}
+
+	if (evt->sync) {
+		if (k_msgq_put(data->event_msgq, &data->pending_event, K_NO_WAIT) != 0) {
+			LOG_DBG("Could not put input data into queue");
+		}
+	}
+}
+
+static int lvgl_input_init(const struct device *dev)
+{
+	const struct lvgl_input_config *cfg = dev->config;
+	struct lvgl_input_data *data = dev->data;
+
+	if (cfg->input_dev && !device_is_ready(cfg->input_dev)) {
+		LOG_ERR("Input device not ready");
+		return -ENODEV;
+	}
+
+	lv_indev_drv_init(&data->indev_drv);
+
+	/* Translate to lv_indev_type_t accounting for LV_INDEV_TYPE_NONE = 0 */
+	switch (cfg->indev_type) {
+	case BINDING_TYPE_POINTER:
+		data->indev_drv.type = LV_INDEV_TYPE_POINTER;
+		break;
+	case BINDING_TYPE_KEYPAD:
+		data->indev_drv.type = LV_INDEV_TYPE_KEYPAD;
+		break;
+	case BINDING_TYPE_BUTTON:
+		data->indev_drv.type = LV_INDEV_TYPE_BUTTON;
+		break;
+	case BINDING_TYPE_ENCODER:
+		data->indev_drv.type = LV_INDEV_TYPE_ENCODER;
+		break;
+	}
+
+	data->indev_drv.read_cb = lvgl_input_read_cb;
+	data->indev = lv_indev_drv_register(&data->indev_drv);
+	data->indev_drv.user_data = dev;
+
+	if (data->indev == NULL) {
+		LOG_ERR("Could not register input device driver");
+		return -EPERM;
+	}
+
+	return 0;
+}
+
+/* The lvgl module intializes its dependencies (memory, display driver) with
+ * CONFIG_APPLICATION_INIT_PRIORITY. To successfully register drivers, the heap has to be
+ * initialized. Also since this device relies on the input subsystem, ensure that this one is ready
+ * too.
+ */
+BUILD_ASSERT(CONFIG_APPLICATION_INIT_PRIORITY < CONFIG_INPUT_LVGL_INIT_PRIORITY);
+BUILD_ASSERT(CONFIG_INPUT_INIT_PRIORITY < CONFIG_INPUT_LVGL_INIT_PRIORITY);
+
+#define INPUT_LVGL_MSGQ_SIZE(inst) (DT_INST_PROP(inst, msgq_size))
+
+#define ASSERT_PROPS(inst)                                                                         \
+	BUILD_ASSERT(!DT_INST_PROP_OR(inst, swap_xy, 0) ||                                         \
+			     DT_INST_ENUM_IDX(inst, type) == BINDING_TYPE_POINTER,                 \
+		     "Property swap-xy only supported for pointer type.");                         \
+	BUILD_ASSERT(!DT_INST_PROP_OR(inst, invert_x, 0) ||                                        \
+			     DT_INST_ENUM_IDX(inst, type) == BINDING_TYPE_POINTER,                 \
+		     "Property invert-x only supported for pointer type.");                        \
+	BUILD_ASSERT(!DT_INST_PROP_OR(inst, invert_y, 0) ||                                        \
+			     DT_INST_ENUM_IDX(inst, type) == BINDING_TYPE_POINTER,                 \
+		     "Property invert-y only supported for pointer type.")
```

Comment:
I think you'd be better off with a binding per type (like a `-common.yaml` file and then an `-pointer.yaml`...), then you'd get these checks automatically from the binding. Same for the config and data structures, you could have a common and specific one, it'd make the whole thing more compact memory wise.

Reply:
I implemented your suggested changes. You were correct, this feels cleaner now :+1: